cargo-near/src/contract.rs

```diff
@@ -0,0 +1,93 @@
+use std::env;
+use std::ffi::OsStr;
+use std::io::BufRead;
+use std::process::Command;
+
+use anyhow::Context;
+
+use crate::cargo::{manifest::CargoManifestPath, metadata::CrateMetadata};
+
+use super::{abi, util, BuildCommand};
+
+fn invoke_rustup<I, S>(args: I) -> anyhow::Result<Vec<u8>>
+where
+    I: IntoIterator<Item = S>,
+    S: AsRef<OsStr>,
+{
+    let rustup = env::var("RUSTUP").unwrap_or_else(|_| "rustup".to_string());
+
+    let mut cmd = Command::new(rustup);
+    cmd.args(args);
+
+    log::info!("Invoking rustup: {:?}", cmd);
+
+    let child = cmd
+        .stdout(std::process::Stdio::piped())
+        .spawn()
+        .context(format!("Error executing `{:?}`", cmd))?;
+
+    let output = child.wait_with_output()?;
+    if output.status.success() {
+        Ok(output.stdout)
+    } else {
+        anyhow::bail!(
+            "`{:?}` failed with exit code: {:?}",
+            cmd,
+            output.status.code()
+        );
+    }
+}
+
+const COMPILATION_TARGET: &str = "wasm32-unknown-unknown";
+
+pub(crate) fn build(args: BuildCommand) -> anyhow::Result<()> {
+    if !invoke_rustup(&["target", "list", "--installed"])?
+        .lines()
+        .any(|target| target.as_ref().map_or(false, |t| t == COMPILATION_TARGET))
+    {
+        anyhow::bail!("rust target `{}` is not installed", COMPILATION_TARGET);
+    }
+
+    let crate_metadata = CrateMetadata::collect(CargoManifestPath::try_from(
+        args.manifest_path.unwrap_or_else(|| "Cargo.toml".into()),
+    )?)?;
```

Comment:
In this case, I think I prefer having this check outside of clap. Primarily because when reporting, clap prefixes the report with “Invalid value” which in most cases, would be a misnomer. As it won't be an invalid value, it just doesn't exist:

```console
error: Invalid value "a/Cargo.toml" for '--manifest-path <PATH>': manifest path does not exist
```

whereas, leaving that as-is, we can have errors like this:

```console
error: manifest path `a/Cargo.toml` does not exist
```

Reply:
I agree, this does look kind of weird. I wouldn't mind having this checked outside of clap.